SchemaMigration/COSchemaMigrationDriver.m

```diff
@@ -42,28 +42,81 @@ static inline void addObjectForKey(NSMutableDictionary *dict, id object, NSStrin
+	return versionsByDomainByEntityTuple[key];
+}
+
+
+
 - (NSSet *)domainsToMigrateForItem: (COItem *)item
 {
-	NSArray *domains = [item valueForAttribute: kCOObjectDomainsProperty];
 	ETEntityDescription *entity = [_modelDescriptionRepository descriptionForName: item.entityName];
 	BOOL isDeletedEntity = (entity == nil);
 	
 	/* For a deleted entity, the domain versions match between item and packages */
 	if (isDeletedEntity)
 	{
-		/* The first domain is the one owning the entity, the remaining domains
-		   own the parent entities. */
-		return S([domains firstObject]);
+		// FIXME: We hit this line when loading a CODictionary during the test suite,
+		// something must be wrong (but no tests fail).
+		return S(item.packageName);
+	}
+	
+	/* Early exit, the item has no version specified, or no package, (i.e.
+	   saved with an old version of CO) so we can't do any migration. */
+	if (item.entityVersion == -1 || item.packageName == nil)
+	{
+		return [NSSet set];
+	}
+	
+	/* Early exit, common case: the version in the item matches the package verion
+	   in the model description repository. In that case we have no migration to do */
+	if (entity.owner != nil
+		&& entity.owner.version == item.entityVersion)
+	{
+		return [NSSet set];
 	}
+	
+	/* At this point we are doing a migration for some domains for sure. */
 
+	NSDictionary *versionsByDomain = [self versionsByDomainForItem: item];
+	
+	if (versionsByDomain == nil
+		|| versionsByDomain[@"org.etoile-project.CoreObject"] == nil
+		|| ![versionsByDomain[item.packageName] isEqual: @(item.entityVersion)])
+	{
+		// TODO: Test that we get this exception when forgetting to call  +[COSchemaMigration recordVersionsByDomain:...]
+		[NSException raise: NSInternalInconsistencyException
+					format: @"Item with entityName '%@' version %d needs a migration, "
+							"but -versionsByDomainForItem: returned an incomplete "
+							"snapshot of the past version of the metamodel we need. "
+							"It returned: %@. \n"
+							"We require it to be a non-nil dictionary, have a "
+							"version set for the org.etoile-project.CoreObject "
+							"package, and include the same package/version as "
+							"the item being migrated. Probably, you forgot to "
+							"call +[COSchemaMigration recordVersionsByDomain:...]",
+							item.entityName, (int)item.entityVersion, versionsByDomain];
+	}
+	
 	NSMutableSet *domainsToMigrate = [NSMutableSet new];
-
-	for (NSString *domain in domains)
+	
+	for (NSString *domain in [versionsByDomain allKeys])
 	{
 		ETPackageDescription *package = [_modelDescriptionRepository descriptionForName: domain];
 		BOOL isDeletedPackage = (package == nil);
+		int64_t version = versionsByDomain[domain] != nil
+			? (int64_t)[versionsByDomain[domain] longLongValue]
+			: (int64_t)-1;
 
-		if (isDeletedPackage || [item versionForDomain: domain] != (int64_t)package.version)
+		if (isDeletedPackage || version != (int64_t)package.version)
```

Comment:
Related to my previous comment, this isDeletedPackage case would have to be turned into an error.